classad/_base_expression.py

```diff
@@ -0,0 +1,119 @@
+import pyparsing as pp
+
+from typing import Iterable, Any, TYPE_CHECKING, Union
+
+if TYPE_CHECKING:
+    from ._expression import ClassAd
+    from ._primitives import Undefined, Error, HTCBool
+
+
+class Expression:
+    _expression: "Expression"
+
+    def evaluate(
+        self, key: Iterable = None, my: "ClassAd" = None, target: "ClassAd" = None
+    ) -> Any:
+        if isinstance(key, str):
+            key = key.split(".")
+        return self._evaluate(key=key, my=my, target=target)
+
+    def _evaluate(
+        self, key: Iterable = None, my: "ClassAd" = None, target: "ClassAd" = None
+    ) -> Any:
+        return NotImplemented
+
+    @classmethod
+    def from_grammar(cls, tokens):
+        result = cls()
+        if isinstance(tokens, pp.ParseResults):
+            # return all primitives as is
+            if len(tokens) == 1:
+                return tokens[0]
+            else:
+                expression = []
+                for token in tokens:
+                    if isinstance(token, pp.ParseResults):
+                        expression.append(token[0])
+                    else:
+                        expression.append(token)
+                result._expression = tuple(expression)
```

Comment:
I would have expected ``result = cls(expression)``. Consider whether you actually need this generic converter in the first place, though.